lib/charms/opensearch/v0/opensearch_relation_provider.py

```diff
@@ -433,3 +435,70 @@ def update_dashboards_password(self):
         pwd = self.secrets.get(Scope.APP, self.secrets.password_key(KibanaserverUser))
         for relation in self.dashboards_relations:
             self.opensearch_provides.set_credentials(relation.id, KibanaserverUser, pwd)
+
+    def _put_relation_user(
+        self, user: str, permissions: dict[str], hashed_pwd: str, relation_id: int
+    ):
+        """Create a relation user.
+
+        Relation users are registered with a dedicated role which maps to the username,
+        and their name is saved in the databag for later reference.
+        """
+        self.user_manager.create_role(role_name=user, permissions=permissions)
+        if not (users := self.charm.peers_data.get_object(Scope.APP, ClientUsersDict)):
+            users = {}
+
+        if users.get(relation_id):
+            logger.warning(
+                "User %s is already registered in Peer Relation data for relation %d.",
+                user,
+                relation_id,
+            )
+
+        self.user_manager.create_user(user, [user], hashed_pwd)
+        users[str(relation_id)] = user
+        self.charm.peers_data.put_object(Scope.APP, ClientUsersDict, users)
+
+    def remove_lingering_relation_users_and_roles(  # noqa: C901
+        self, departed_relation_id: int | None = None
+    ):
+        """Removes lingering relation users and roles from opensearch.
+
+        Args:
+            departed_relation_id: if a relation is departing, pass in the ID and its user will be
+                deleted.
+        """
+        if not self.opensearch.is_node_up() or not self.unit.is_leader():
+            return
+
+        if not (relation_users := self.charm.peers_data.get_object(Scope.APP, ClientUsersDict)):
+            relation_users = {}
+
+            if departed_relation_id:
+                logging.warning(
+                    "User for relation %d wasn't registered in internal cham workflows.",
+                    departed_relation_id,
+                )
+
+        clearnup_rel_ids = []
+        if departed_relation_id:
+            clearnup_rel_ids = [str(departed_relation_id)]
+        else:
+            rel_ids = [str(relation.id) for relation in self.opensearch_provides.relations]
+            clearnup_rel_ids = list(set(relation_users.keys()) - set(rel_ids))
+
+        for rel_id in clearnup_rel_ids:
+            if username := relation_users.get(rel_id):
```

Comment:
Could you explain when would this be None?
the `del` below suggests otherwise (it will raise a key error if it was ever None)

Reply:
With the current code It could be `None` if --for example-- for any "unexpected" reason the function is called with a `departed_relation_id` for a relation that may not have been registered in the databag (anymore).

I'm adding changes so that this will not be an option anymore.

Yet, TBH I find it safer to ensure smooth user experience over non-crtiical errors. We had too many issues (https://github.com/canonical/opensearch-operator/issues/323, https://github.com/canonical/opensearch-operator/issues/222, https://github.com/canonical/opensearch-operator/issues/422) on being confident referencing internal data structures with no security measures.
(I agree that we should signal any inconsistency, but I wonder if it's worth to put the charm in an `error` state. Ideally the internal data structure should recover itself, and the `update-status` usage ensures no lingering users staying in the system.)

You are right that the `del` below is incorrect without a `try...catch`. Changing it so we log and exit in case the key was not found. 

Shame on me not to have had a unittest on the corner case. Adding that, too.

Reply:
Done. (Tests covered the refactored code.)